src/Data/Singletons/Single/Type.hs

```diff
@@ -89,3 +113,184 @@ singPredRec _ctx DArrowT =
+signature, we could get away without an explicit `forall` in the singled type
+signature. That is, one could write:
+
+  sAbsurd :: Sing (v :: V a) -> Sing (Absurd :: b)
+
+This would have the right type variable order, but unfortunately, this approach
+does not play well with singletons' style of code generation. Consider the code
+that would be generated for the body of sAbsurd:
+
+  sAbsurd :: Sing (v :: V a) -> Sing (Absurd :: b)
+  sAbsurd (sV :: Sing v) = id @(Case v v :: b) (case sV of {})
+
+Note the use of the type `Case v v :: b` in the right-hand side of sAbsurd.
+However, because `b` was not bound by a top-level `forall`, it won't be in
+scope here, resulting in an error!
+
+(Why do we generate the code `id @(Case v v :: b)` in the first place? See
+Note [The id hack; or, how singletons learned to stop worrying and avoid kind generalization]
+in D.S.Single.)
+
+The simplest approach is to just always generate singled type signatures with
+explicit `forall`s. In the event that the original type signature lacks an
+explicit `forall`, we inferr the correct type variable ordering ourselves and
+synthesize a `forall` with that order. The `singTypeKVBs` function implements
+this logic.
+
+-----
+-- Wrinkle 2: The TH reification swamp
+-----
+
+There is another issue with type signatures that lack explicit `forall`s, one
+which the current design of Template Haskell does not make simple to fix.
+If we single code that is wrapped in TH quotes, such as in the following example:
+
+  $(singletons [d|
+    data Proxy (a :: k) where
+      MkProxy :: Proxy a
+    |])
+
+Then our job is made much easier when singling MkProxy, since we know that the
+only type variable that must be quantified is `a`, as that is the only one
+specified in the type signature.
+
+However, this is not the only possible way to single MkProxy. One can
+alternatively use $(genSingletons [''Proxy]), which uses TH reification to
+infer the type of MkProxy. There is perilous, however, because this is how
+TH reifies MkProxy:
+
+  ForallC [KindedTV k StarT,KindedTV a (VarT k)] []
+          (GadtC [MkProxy] [] (AppT (ConT Proxy) (VarT a)))
+
+In terms of actual Haskell code, that's:
+
+  MkProxy :: forall k (a :: k). Proxy a
+
+This is subtly different than before, as `k` is now specified. Contrast this
+with `MkProxy :: Proxy a`, where `k` is invisible. In other words, if you
+single MkProxy using genSingletons, then `Proxy @True` will typecheck but
+`SMkProxy @True` will /not/ typecheck—you'd have to use `SMkProxy @_ @True`
+instead. Urk!
+
+At present, Template Haskell does not have a way to distinguish specified from
+inferred type variables, so it is unclear how one could work around this issue.
+Thankfully, it is only likely to bite in situations where the original type
+signature uses inferred variables, so the damage is somewhat minimal.
+
+-----
+-- Wrinkle 3: Where to put explicit kind annotations
```

Comment:
Where does this wrinkle manifest itself in code?

Reply:
I've added some forward references to this wrinkle in the code.